ISPN-12438 off heap details (#8803)
infinispan/infinispan: documentation/src/main/asciidoc/topics/con_off_heap_memory.adoc

Consider the following simplified illustration of memory space for a JVM process where {brandname} is running:

image::offheap.svg[]

.JVM heap memory
The heap is divided into young and old generations that help keep referenced Java objects and other application data in memory.
The Garbage Collector (GC) process reclaims space from unreachable objects, running more frequently on the young generation memory pool.

When {brandname} stores cache entries in JVM heap memory, GC runs can take longer to complete as you start adding data to your caches.
Because GC is a fairly intensive process, longer and more frequent runs can degrade application performance.

.Off heap memory
Off heap memory is native available system memory outside JVM memory management.
The preceding diagram shows the `Metaspace` memory pool that holds class metadata and is allocated from native memory.
The diagram also represents a section of native memory that holds {brandname} cache entries.

.Storing data off heap
When you add entries to off heap caches, {brandname} dynamically allocates native memory to your data.

> {brandname} hashes the serialized `byte[]` for each key into buckets that are similar to a standard Java `HashMap`.

Reviewer: We should probably mention that this means that equality is done via the serialized bytes and not the object representation as well. I believe we had that somewhere before... but don't remember where.

Author: @wburns Please check the note I added about equality: https://github.com/infinispan/infinispan/pull/8803/files#diff-b632766944b1fc30eb18b540ff43965eb166f1f10211f64cab65b295daa5a70eR39

Buckets include address pointers that {brandname} use to locate entries that you store in off heap memory.

The following diagram shows a set of keys with names, the hash for each key and bucket array of address pointers, as well as the entries with the name and phone number:

image::offheap_hashmap.svg[]

In cases where key hashes collide, {brandname} links entries.
As in the preceding diagram, if the William Clay and Luke Cage keys have the same hash, then the first entry added to the cache is the first element in the bucket.

[NOTE]
====
Even though {brandname} stores cache entries in native memory, run-time operations require JVM heap representations of those objects.
For instance, `cache.get()` operations read objects into heap memory before returning.
Likewise, state transfer operations hold subsets of objects in heap memory while they take place.
====

.Memory overhead

Memory overhead is the additional memory that {brandname} uses to store entries.

> {brandname} uses 25 bytes for each entry entry in the cache.

Reviewer 2: s/entry entry/entry/

> Using eviction to create a bounded off heap data container increases memory overhead by 36 bytes.

Reviewer 2: I would not leave the total implied. You are mentioning it below, but it's clearer if brought here,

This happens because {brandname} creates an additional linked list to track entries in the cache and perform eviction.

> However, compared with storing entries in the JVM heap, a total memory overhead of 61 bytes per entry is significantly lower.

Reviewer 2: @wburns can we provide a rough estimation of on heap overhead ?

Author: @wburns I agree that it would be good, if possible, to provide a rough estimation of on-heap. That way we can back up the statement that off-heap overhead is much less.

[id='off_heap_memory']
= Storing Data in Off Heap Memory
{brandname} can store cache entries in JVM heap memory or off heap.

> Off heap storage lets you use unmanaged, native memory space for the JVM.

Reviewer: ```suggestion
Off heap storage lets you use unmanaged, native memory space for the JVM.
```
Off heap storage lets your objects occupy native memory outside of the managed JVM memory.

> Off heap storage gives you less overhead per entry than in JVM heap memory.

Reviewer: Gives sounds a bit weird. Maybe "requires less overhead per entry" ?

In other words, off heap storage uses less memory than heap storage for the same amount of data.
Another benefit of off heap storage is that it is not affected by Garbage Collector runs, which can improve overall JVM performance.

> Likewise you cannot change the JVM heap space at run-time but, when using off heap storage, {brandname} dynamically allocates memory.

Reviewer: ```suggestion
Likewise you cannot change the JVM heap space at run-time, but when using off heap storage, {brandname} dynamically allocates memory for each entry.
```

> However there are some trade-offs such as that heap dumps become worthless.

Reviewer: ```suggestion
However there are some trade-offs such as JVM heap dumps will now show entries stored in an off heap cache.
```

Consider the following illustration of memory space for a JVM process where
{brandname} is running:

image::offheap.svg[]

JVM heap memory, as shown in the preceding diagram, holds Java objects and other application data.
The heap is divided into young and old generations that help keep referenced objects in memory.
The Garbage Collector (GC) process reclaims space from unreachable objects, running more frequently on the young generation memory pool.

When {brandname} stores cache entries in JVM heap memory, GC runs can take longer to complete as you start adding data to your caches.
Because GC is a fairly intensive process, longer and more frequent runs can degrade application performance.

Using off heap memory configures {brandname} to store data in native memory, or available system memory that is outside JVM memory management.
Depending on your JVM implementation, native memory is also allocated for thread stacks, Just-in-Time (JIT) compiler code, and so on.
As an example, the preceding diagram shows two memory pools that are allocated from native memory:

* Metaspace holds class metadata.

> * Code Cache holds generated native code and internal structures.

Reviewer: We may want to rename this something different as the Code native region is not where objects allocated via Infinispan.

Author: indeed. I'm reworking this part. I thought that's what I was observing when adding to an off heap cache through the Java console.

Author: What if we leave it as-is in the diagram as a basic representation of what we mean when we say "off heap"? 

Here `Code Cache` is just one type of un-managed memory pool.

> Taking the diagram as an example, when you use off heap storage {brandname} adds data to the Code Cache memory pool.

Reviewer: Technically they are allocated in the "Unknown" region as shown by native memory tracking.

Author: I've updated to keep the wording a little more general. "Infinispan allocates native memory to data" is probably the thing we should say rather than mentioning specific memory pools.

When you add an entry to a cache that uses off heap storage, {brandname} hashes the serialized `byte[]` for the key into buckets that are similar to a standard Java `HashMap`.
Buckets include address pointers that {brandname} use to locate entries in off heap memory.

The following diagram shows a set of keys with names, the hash for each key and bucket array of address pointers, as well as the entries with the name and phone number:

image::offheap_hashmap.svg[]

> In cases where key hashes collide, {brandname} chains entries.

Reviewer: maybe links instead of chains? It is really a linked list.

As in the preceding diagram, if the William Clay and Luke Cage keys have the same hash, then the first entry added to the cache is the first element in the bucket.

.Memory overhead

Memory overhead is the additional memory that {brandname} uses to store entries.
{brandname} uses 25 bytes for each entry entry in the cache.

Using eviction to create a bounded off heap data container increases memory overhead by 36 bytes.
This happens because {brandname} creates an additional linked list to track entries in the cache and perform eviction.

However, compared with storing entries in the JVM heap, a total memory overhead of 61 bytes per entry is significantly lower.

.Data consistency

> {brandname} uses an array of `ReadWriteLock` instances to protect off heap address spaces.

Reviewer: Technically we use StampedLock now, but you can think of it as a ReadWriteLock.

Author: how about if we just say locks as a more generic thing?

The number of `ReadWriteLock` instances is twice the number of cores and then rounded to the nearest power of two.

> This ensures that there is an even distribution of `ReadWriteLock` instances to prevent read operations from blocking each other.

Reviewer: The distribution of read locks is not so they wouldn't block each other as a the read portion of a read write lock will not block another read. The distribution is to prevent reads from being blocked by a write lock.

However write operations wait to complete, which blocks read operations.

[NOTE]
====

> {brandname} requires JVM heap memory for normal operations. For instance, state

Reviewer: I wonder if this should be more like `While entries are stored in native memory, operations require using a JVM heap representation of those objects. For instance, a get operation will read the object into native memory before returning. Likewise, state transfer will hold a subset of objects in memory while transferring them another node."

Something like that maybe?

Author: nice. I've updated with that.
